### PR TITLE
Minor refactoring

### PR DESCRIPTION
**Pull request by mortenhauberg**

### CODE REVIEW

File: src/Client.php [hunk 1]
```diff
@@ -190,13 +190,10 @@ private function configureDefaults(array $config)
...
-            // Add the User-Agent header if one was not already set.
-            foreach (array_keys($this->config['headers']) as $name) {
-                if (strtolower($name) === 'user-agent') {
-                    return;
-                }
-            }
+        }
+
+        $lowercasedHeaders = array_change_key_case($this->config['headers'], CASE_LOWER);
+        if (!isset($lowercasedHeaders['user-agent'])) {
```
Comment (sagikazarmark) [inline src/Client.php]: I would still leave this in an else branch. This now runs nevertheless of the previous if, while it should only run if the previous condition is false.

File: src/Handler/CurlFactory.php [hunk 2]
```diff
@@ -227,12 +227,13 @@ private function applyMethod(EasyHandle $easy, array &$conf)
         }
 
         $method = $easy->request->getMethod();
-        if ($method === 'PUT' || $method === 'POST') {
+        if (in_array($method, ['PUT', 'POST'], true)
```
Comment (sagikazarmark) [inline src/Handler/CurlFactory.php]: I don't think this increases readability. What's the reasoning behind?

File: src/UriTemplate.php [hunk 3]
```diff
@@ -131,9 +131,7 @@ private function expandMatch(array $matches)
 
                     if (!$isNestedArray) {
                         $var = rawurlencode($var);
-                        if ($parsed['operator'] === '+' ||
-                            $parsed['operator'] === '#'
-                        ) {
+                        if (in_array($parsed['operator'], ['+', '#'], true)) {
```
Comment (sagikazarmark) [inline src/UriTemplate.php]: Same here: don't think it is better readable, even if shorter.

File: src/UriTemplate.php [hunk 4]
```diff
@@ -185,7 +183,7 @@ private function expandMatch(array $matches)
                     $variable = substr($variable, 0, $value['position']);
                 }
                 $expanded = rawurlencode($variable);
-                if ($parsed['operator'] === '+' || $parsed['operator'] === '#') {
+                if (in_array($parsed['operator'], ['+', '#'], true)) {
```
Comment (sagikazarmark) [inline src/UriTemplate.php]: And here too

File: src/Client.php (hunk 1 above)
Comment (mortenhauberg) [inline src/Client.php]: 👍🏼

File: src/Handler/CurlFactory.php (hunk 2 above)
Comment (mortenhauberg) [inline src/Handler/CurlFactory.php]: In my mind it does increase the readability, but I'll change it :)

File: src/UriTemplate.php (hunk 3 above)
Comment (mortenhauberg) [inline src/UriTemplate.php]: I'll revert :)

File: src/UriTemplate.php (hunk 4 above)
Comment (mortenhauberg) [inline src/UriTemplate.php]: 👍🏼

File: src/Handler/CurlFactory.php (hunk 2 above)
Comment (sagikazarmark) [inline src/Handler/CurlFactory.php]: Option one: "If the method is PUT or POST then..."
Option two: "if the method is one of PUT or POST then..."

For me the first one is logically more simple as we want a literally logical answer to this question. That's how I read it. :)

File: src/Handler/CurlFactory.php [hunk 5]
```diff
@@ -227,12 +227,12 @@ private function applyMethod(EasyHandle $easy, array &$conf)
...
+        if (($method === 'PUT' || $method === 'POST') && !$easy->request->hasHeader('Content-Length')) {
             // See http://tools.ietf.org/html/rfc7230#section-3.3.2
-            if (!$easy->request->hasHeader('Content-Length')) {
-                $conf[CURLOPT_HTTPHEADER][] = 'Content-Length: 0';
-            }
-        } elseif ($method === 'HEAD') {
+            $conf[CURLOPT_HTTPHEADER][] = 'Content-Length: 0';
+        }
+
+        if ($method === 'HEAD') {
```
Comment (sagikazarmark) [inline src/Handler/CurlFactory.php]: I am still confused about this: If the method is PUT or POST, I don't want to evaluate this second condition, so elseif is better. But in that case, this needs to be reverted too. Sorry.

Comment (mortenhauberg) [inline src/Handler/CurlFactory.php]: Sure 👍🏼

File: src/Handler/CurlFactory.php [hunk 6]
```diff
@@ -227,11 +227,9 @@ private function applyMethod(EasyHandle $easy, array &$conf)
         }
 
         $method = $easy->request->getMethod();
-        if ($method === 'PUT' || $method === 'POST') {
+        if (($method === 'PUT' || $method === 'POST') && !$easy->request->hasHeader('Content-Length')) {
```
Comment (sagikazarmark) [inline src/Handler/CurlFactory.php]: Sorry, I meant this thing too. Right now if the method is PUT or POST, but the second condition evaluates to false the elseif condition is still evaluated.

Comment (mortenhauberg) [inline src/Handler/CurlFactory.php]: You're totally right

**Comment by sagikazarmark** [comment]
Sorry for asking you to revert most of your changes. :(

**Comment by sagikazarmark** [comment]
And thank you very much

**Comment by mortenhauberg** [comment]
No problem. And thank you!

**Comment by gmponos** [comment]
Hello,

What I was wondering about this MR is if these changes have any minor impact in performance. I haven't done any performance test on this MR but from what I know array_filter is slower than foreach and `is_null` is slower from `=== null`

Most of the changes I saw on this MR where mainly about making the code more readable. I guess for a library like guzzle performance should be most important rather than readability.

sources: 
http://www.levijackson.net/are-array_-functions-faster-than-loops/
http://stackoverflow.com/a/4662685/4158811

Probably you can find other sources too.

**Comment by sagikazarmark** [comment]
Generally I agree with you. The reason why I accepted this patch is that (correct me if I am wrong) these are only called upon construction time once, so it should not cause any performance hit.

**Comment by gmponos** [comment]
Hmm.. I didn't notice that. Yes you are correct.. It is called only on construction... So yes, most probably the difference will be insignificance.

**Comment by sagikazarmark** [comment]
Actually I was wrong in one of the cases: it's called in every subsequent request. It's premature optimization anyway, so I am reverting the PR.

**Pull request merged**